KAFKA-4218: Add withkey methods to KStream

jeyhunkarimov opened the pull request:
This PR aims to add withKey methods to KStream interface. 
This PR is part of [KIP-149](https://cwiki.apache.org/confluence/display/KAFKA/KIP-149%3A+Enabling+key+access+in+ValueTransformer%2C+ValueMapper%2C+and+ValueJoiner).
I separated the complete PR into 4 parts as discussed in [here](https://github.com/apache/kafka/pull/3570#issuecomment-317645747). So, this PR concentrates on adding withKey methods to KStream interface.

jeyhunkarimov committed:
Add withkey interfaces

Add withkey interfaces

Add withKey converters in AbstractStream

Add KStream withKey interfaces

Remove unnecessary lines

Fix typos, replace joiner/mapper-withKey to mapper/joiner

Add JoinerWithKey tests

Rebase KStreamImplTest.java

FIx numProcesTest()

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6435/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6420/
Test PASSed (JDK 8 and Scala 2.12).

dguy commented:
@bbejeck @mjsax @enothereska @guozhangwang reviews please

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/InitializerWithKey.java:
```diff
@@ -0,0 +1,42 @@
+package org.apache.kafka.streams.kstream;
+
+/**
+ * The {@code InitializerWithKey} interface for creating an initial value in aggregations with read-only key.
```
"with a read-only key"

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/InitializerWithKey.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public interface InitializerWithKey<K, VA> {
+    /**
+     * Return the initial value for an aggregation given read-only key.
```
"given a read-only key"

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -180,6 +180,41 @@
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KStream<String, String> inputStream = builder.stream("topic");
+     * KStream<String, Integer> outputStream = inputStream.mapValues(new ValueMapper<String, Integer> {
```
This should be new `ValueMapperWithKey`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -180,6 +180,41 @@
+     * @see #selectKey(KeyValueMapper)
+     * @see #map(KeyValueMapper)
+     * @see #flatMap(KeyValueMapper)
+     * @see #flatMapValues(ValueMapper)
```
also, `flatMapValue(ValueMapperWithKey)

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -264,6 +299,47 @@
+     * The example below splits input records {@code <null:String>} containing sentences as values into their words.
+     * <pre>{@code
+     * KStream<byte[], String> inputStream = builder.stream("topic");
+     * KStream<byte[], String> outputStream = inputStream.flatMapValues(new ValueMapper<String, Iterable<String>> {
```
`ValueMapperWithKey`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -264,6 +299,47 @@
+     * Thus, <em>no</em> internal data redistribution is required if a key based operator (like an aggregation or join)
+     * is applied to the result {@code KStream}. (cf. {@link #flatMap(KeyValueMapper)})
+     *
+     * @param processor a {@link ValueMapperWithKey} the computes the new output value with read-only key
```
Can probably remove the "with read-only key" bit here and elsewhere as it is already mentioned in the comment above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -264,6 +299,47 @@
+     * @see #selectKey(KeyValueMapper)
+     * @see #map(KeyValueMapper)
+     * @see #flatMap(KeyValueMapper)
+     * @see #mapValues(ValueMapper)
```
`mapValue(ValueMapperWithKey)` ?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -942,6 +1018,83 @@ void to(final Serde<K> keySerde,
+     * pairs should be emitted via {@link ProcessorContext#forward(Object, Object)
+     * ProcessorContext.forward()}.
+     * <pre>{@code
+     * new ValueTransformerSupplier() {
```
`ValueTransformerWithKeySupplier`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1302,7 +1445,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * <td>&lt;K1:A&gt;</td>
      * <td></td>
-     * <td>&lt;K1:ValueJoiner(A,null)&gt;</td>
+     * <td></td>
```
Not sure this change was deliberate?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1381,7 +1530,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * <td>&lt;K1:A&gt;</td>
      * <td></td>
-     * <td>&lt;K1:ValueJoiner(A,null)&gt;</td>
+     * <td></td>
```
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1477,12 +1623,12 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * <td></td>
      * <td>&lt;K3:c&gt;</td>
-     * <td>&lt;K3:ValueJoiner(null,c)&gt;</td>
+     * <td></td>
```
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1927,6 +2827,91 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+     * @param <VT>          the value type of the table
+     * @param <VR>          the value type of the result stream
+     * @return a {@code KStream} that contains join-records for each key and values computed by the given
+     * {@link ValueJoiner}, one output for each input {@code KStream} record
```
`ValueJoinerWithKey`

dguy replied:
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1534,42 +2195,286 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+     * @param <VT>          the value type of the table
+     * @param <VR>          the value type of the result stream
+     * @return a {@code KStream} that contains join-records for each key and values computed by the given
+     * {@link ValueJoiner}, one for each matched record-pair with the same key
```
`ValueJoinerWithKey`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1832,17 +2732,17 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+     * @param <VT>          the value type of the table
+     * @param <VR>          the value type of the result stream
      * @return a {@code KStream} that contains join-records for each key and values computed by the given
      * {@link ValueJoiner}, one output for each input {@code KStream} record
```
`ValuejoinerWithKey`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1958,6 +2943,39 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+     * @param <GV>           the value type of the {@link GlobalKTable}
+     * @param <RV>           the value type of the resulting {@code KStream}
+     * @return a {@code KStream} that contains join-records for each key and values computed by the given
+     * {@link ValueJoiner}, one output for each input {@code KStream} record
```
`ValueJoinerWithKey`

dguy replied:
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1992,4 +3010,40 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+     * @param <GV>               the value type of the {@link GlobalKTable}
+     * @param <RV>               the value type of the resulting {@code KStream}
+     * @return a {@code KStream} that contains join-records for each key and values computed by the given
+     * {@link ValueJoiner}, one output for each input {@code KStream} record
```
`ValueJoinerWithKey`

dguy replied:
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ReducerWithKey.java:
```diff
@@ -0,0 +1,53 @@
+import org.apache.kafka.streams.KeyValue;
+
+/**
+ * The {@code ReducerWithKey} interface for combining two values of the same type into a new value with read-only key.
```
probably remove "with read-only key"

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueJoinerWithKey.java:
```diff
@@ -0,0 +1,57 @@
+package org.apache.kafka.streams.kstream;
+
+/**
+ * The {@code ValueJoinerWithKey} interface for joining two values into a new value of arbitrary type given read-only
```
remove "given read-only key"

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapperWithKey.java:
```diff
@@ -0,0 +1,49 @@
+package org.apache.kafka.streams.kstream;
+
+/**
+ * The {@code ValueMapperWithKey} interface for mapping a value to a new value of arbitrary type given read-only key.
```
same here

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,119 @@
+ * This is a stateful record-by-record operation, i.e, {@link #transform(Object, Object)} is invoked individually for
+ * each record of a stream and can access and modify a state that is available beyond a single call of
+ * {@link #transform(Object, Object)} (cf. {@link ValueMapperWithKey} for stateless value transformation).
+ * Additionally, this {@code ValueTransformer} can {@link ProcessorContext#schedule(long, PunctuationType, Punctuator) schedule}
```
`ValueTransformerWithKey` - here and elsewhere

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,119 @@
+ * a method to be {@link Punctuator#punctuate(long) called periodically} with the provided context.
+ * If {@code ValueTransformer} is applied to a {@link KeyValue} pair record the record's key is preserved.
+ * <p>
+ * Use {@link ValueTransformerSupplier} to provide new instances of {@code ValueTransformer} to Kafka Stream's runtime.
```
`ValueTransformerWithKeySupplier`?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,119 @@
+     * Note that using {@link ProcessorContext#forward(Object, Object)},
+     * {@link ProcessorContext#forward(Object, Object, int)}, or
+     * {@link ProcessorContext#forward(Object, Object, String)} is not allowed within any method of
+     * {@code ValueTransformer} and will result in an {@link StreamsException exception}.
```
`ValueTransformerWithKey`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Transform the given value to a new value. {@code key} is read-only and should not be modified. Any modification
+     * to {@code key} can result in corrupt partitioning.
+     * Additionally, any {@link StateStore} that is {@link KStream#transformValues(ValueTransformerSupplier, String...)
```
`ValueTransformerWithKeySupplier` ?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -112,5 +133,78 @@ public R apply(T2 value2, T1 value1) {
+            }
+
+            @Override
+            public VR punctuate(long timestamp) {
```
`@SupressWarnings("deprecation")`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -112,5 +133,78 @@ public R apply(T2 value2, T1 value1) {
     }
 
+    static <K, V, VR> ValueMapperWithKey<K, V, VR> withKey(final ValueMapper<V, VR> valueMapper) {
+        Objects.requireNonNull(valueMapper);
```
add a message here and elsewhere, i.e, "valueMapper can't be null" ?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -112,5 +133,78 @@ public R apply(T2 value2, T1 value1) {
+        };
+    }
+
+    static <K, V> ReducerWithKey<K, V> withKey(final Reducer<V> reducer) {
```
The next 2 methods aren't used in this PR. I guess they are used in one of the others...

dguy replied:
Can we remove these and add them in one of the other PRs where they are used?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValues.java:
```diff
@@ -37,10 +37,10 @@
         @Override
         public void process(K key, V value) {
-            Iterable<? extends V1> newValues = mapper.apply(value);
+            Iterable<? extends V1> newValues = mapperWithKey.apply(key, value);
```
Tests for this?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -27,6 +27,9 @@
 import org.apache.kafka.streams.kstream.KGroupedStream;
 import org.apache.kafka.streams.kstream.GlobalKTable;
 import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.ValueMapperWithKey;
```
nit: group these imports with the other `ValueJoiner` etc imports

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -431,11 +446,18 @@ public void to(final Serde<K> keySerde, final Serde<V> valSerde, StreamPartition
+    }
+
+    @Override
+    public <VR> KStream<K, VR> transformValues(ValueTransformerWithKeySupplier<? super K, ? super V, ? extends VR> valueTransformerWithKeySupplier,
```
`final` params?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -479,7 +517,14 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
+    }
+
+    @Override
+    public <VO, VR> KStream<K, VR> outerJoin(final KStream<K, VO> otherStream,
```
nit: put each param on a separate line?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -583,13 +635,22 @@ public boolean test(final K1 key, final V1 value) {
+    }
+
+    @Override
+    public <V1, R> KStream<K, R> leftJoin(final KStream<K, V1> otherStream,
```
nit: align params

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -598,69 +659,103 @@ public boolean test(final K1 key, final V1 value) {
+
+    @Override
+    public <VO, VR> KStream<K, VR> leftJoin(final KStream<K, VO> otherStream,
+                                            final ValueJoinerWithKey<? super K, ? super V, ? super VO, ? extends VR> joinerWithKey,
```
i think we should just stick with `joiner` for the name of this param. here and elsewhere

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -31,14 +31,14 @@
     private final long joinAfterMs;
 
-    private final ValueJoiner<? super V1, ? super V2, ? extends R> joiner;
+    private final ValueJoinerWithKey<? super K, ? super V1, ? super V2, ? extends R> joinerWithKey;
```
nit: `joiner`?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoin.java:
```diff
@@ -30,18 +30,20 @@ public K apply(final K key, final V1 value) {
     };
     private final KTableValueGetterSupplier<K, V2> valueGetterSupplier;
-    private final ValueJoiner<? super V1, ? super V2, R> joiner;
+    private final ValueJoinerWithKey<? super K, ? super V1, ? super V2, R> joinerWithKey;
```
here too. And elsewhere

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamMapValues.java:
```diff
@@ -16,17 +16,17 @@
 class KStreamMapValues<K, V, V1> implements ProcessorSupplier<K, V> {
 
-    private final ValueMapper<V, V1> mapper;
+    private final ValueMapperWithKey<K, V, V1> mapperWithKey;
```
`mapper`

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -235,7 +238,17 @@ public void shouldNotAllowNullMapperOnMap() throws Exception {
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotAllowNullMapperWothKeyOnMapValues() throws Exception {
```
typo: Woth -> With

dguy reviewed (commented):
Thanks @jeyhunkarimov, I think we need to add some tests for the new interfaces? I don't really see them being used anywhere.

jeyhunkarimov committed:
Change valueJoinerWithKey -> joiner

Move method arguments one line up - cosmetic changes

Remove repetition in tests and remove addProcessor for non-key interfaces

Solve Cyclomatic Complexity error in tests

jeyhunkarimov commented:
@dguy Sorry for super long delay ( I was on vacation) and thanks for review.
I propose first concentrating on this PR and after it is (hopefully) merged, we can incrementally change the others and complete the KIP in 4 RPs. WDYT? Btw, I submitted a patch, cc\ @mjsax @bbejeck @guozhangwang

dguy commented:
Hey @jeyhunkarimov - no probs. Sounds like a good plan. Can @mjsax or @bbejeck please review this patch?

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -472,16 +492,21 @@ public void to(final Serde<K> keySerde,
     @Override
-    public <V1> KStream<K, V1> transformValues(final ValueTransformerSupplier<? super V, ? extends V1> valueTransformerSupplier,
-                                               final String... stateStoreNames) {
+    public <VR> KStream<K, VR> transformValues(final ValueTransformerSupplier<? super V, ? extends VR> valueTransformerSupplier, final String... stateStoreNames) {
```
nit: args on separate line

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -497,31 +522,62 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
-        return doJoin(other, joiner, windows, keySerde, thisValueSerde, otherValueSerde,
-            new KStreamImplJoin(false, false));
+    public <V1, R> KStream<K, R> join(
+            final KStream<K, V1> other,
```
super nit: move method params up one line to start after `join(`

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -497,31 +522,62 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
-                                           final Serde<V1> otherValueSerde) {
-        return doJoin(other, joiner, windows, keySerde, thisValueSerde, otherValueSerde, new KStreamImplJoin(true, true));
+    public <V1, R> KStream<K, R> join(
+        final KStream<K, V1> other,
```
super nit: ditto from above

bbejeck commented:
@jeyhunkarimov made an initial pass, just a few minor comments, will take another look.

jeyhunkarimov commented:
Thanks @bbejeck for the review. I fixed nittos

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/InitializerWithKey.java:
```diff
@@ -0,0 +1,42 @@
+package org.apache.kafka.streams.kstream;
+
+/**
+ * The {@code InitializerWithKey} interface for creating an initial value in aggregations with read-a only key.
```
with a read only key

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1237,7 +1417,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * internal repartitioning topic in Kafka and write and re-read the data via this topic before the actual join.
      * The repartitioning topic will be named "${applicationId}-XXX-repartition", where "applicationId" is
-     * user-specified in {@link StreamsConfig} via parameter
+     * user-specified in {@link  StreamsConfig} via parameter
```
nit: this change doesn't look intentional

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKeySupplier.java:
```diff
@@ -0,0 +1,41 @@
+ * @param <V>  value type
+ * @param <VR> transformed value type
+ * @see ValueTransformerWithKey
+ * TODO
```
?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -180,8 +183,16 @@ public String apply(K key, V value) {
         Objects.requireNonNull(mapper, "mapper can't be null");
         String name = builder.newName(MAPVALUES_NAME);
```
remove this line

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -180,8 +183,16 @@ public String apply(K key, V value) {
         String name = builder.newName(MAPVALUES_NAME);
 
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<>(mapper), this.name);
+        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<>(withKey(mapper)), this.name);
```
we need to remove this line, too. It is taken car of in `mapValues(ValueMapperWithKey...)`

jeyhunkarimov replied:
very nice catch. Thanks.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -319,7 +330,16 @@ public void writeAsText(final String filePath,
         Objects.requireNonNull(mapper, "mapper can't be null");
         String name = builder.newName(FLATMAPVALUES_NAME);
```
remove this line

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -319,7 +330,16 @@ public void writeAsText(final String filePath,
         String name = builder.newName(FLATMAPVALUES_NAME);
 
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamFlatMapValues<>(mapper), this.name);
+        builder.internalTopologyBuilder.addProcessor(name, new KStreamFlatMapValues<>(withKey(mapper)), this.name);
```
remove this line

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -769,13 +894,13 @@ public boolean test(final K1 key, final V1 value) {
-                                                   final Serde<K1> keySerde,
-                                                   final Serde<V1> lhsValueSerde,
-                                                   final Serde<V2> otherValueSerde) {
+        public <K1, R, V1, V2> KStream<K1, R> join(KStream<K1, V1> lhs,
```
can we make the params final?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java:
```diff
@@ -36,29 +36,29 @@
 
-    public KStreamTransformValues(ValueTransformerSupplier<V, R> valueTransformerSupplier) {
-        this.valueTransformerSupplier = valueTransformerSupplier;
+    public KStreamTransformValues(ValueTransformerWithKeySupplier<K, V, R> valueTransformerWithKeySupplier) {
```
nit: `final`
I know it wasn't before, but let's stick to making params as final

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java:
```diff
@@ -36,29 +36,29 @@
 
-        public KStreamTransformValuesProcessor(ValueTransformer<V, R> valueTransformer) {
-            this.valueTransformer = valueTransformer;
+        public KStreamTransformValuesProcessor(ValueTransformerWithKey<K, V, R> valueTransformerWithKey) {
```
ditto

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -454,9 +455,9 @@ public void to(final Serde<K> keySerde,
         String name = builder.newName(TOSTREAM_NAME);
 
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<K, Change<V>, V>(new ValueMapper<Change<V>, V>() {
+        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<K, Change<V>, V>(new ValueMapperWithKey<K, Change<V>, V>() {
```
You can remove the type params from `KStreamMapValues`

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -454,9 +455,9 @@ public void to(final Serde<K> keySerde,
+        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<K, Change<V>, V>(new ValueMapperWithKey<K, Change<V>, V>() {
             @Override
-            public V apply(Change<V> change) {
+            public V apply(K key, Change<V> change) {
```
Can we change these to `final`?

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -82,16 +85,16 @@ public boolean test(String key, String value) {
             });
 
-        KStream<String, Integer> stream2 = stream1.mapValues(new ValueMapper<String, Integer>() {
+        KStream<String, Integer> stream2 = stream1.mapValues(new ValueMapperWithKey<String, String, Integer>() {
```
we need tests with both `ValueMapper` and `ValueMapperWithKey`. Same applies to other methods that have been added

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -156,6 +156,69 @@ public void testJoin() throws Exception {
 
+
+    @Test
+    public void testJoinWithKey() throws Exception {
```
These tests that have been added are largely the same as the non key tests. Is it possible to refactor them so that they share the common code?

jeyhunkarimov replied:
@dguy I tried to apply tests (with and without key) through a common method (like `testJoin(boolean withKey)`), however, I got Cyclomatic complexity error while testing. As a result, I decided to pass some complexity to another function `KStreamKstreamJoinTest.checkResult()`.

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -147,6 +147,64 @@ public void testLeftJoin() throws Exception {
     }
 
     @Test
+    public void testLeftJoinWithKey() throws Exception {
```
Same here

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:
```diff
@@ -133,4 +133,83 @@ public void testJoin() throws Exception {
     }
+
+    @Test
+    public void testJoinWithKey() throws Exception {
```
and here

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableLeftJoinTest.java:
```diff
@@ -134,4 +134,83 @@ public void testJoin() throws Exception {
     }
+
+    @Test
+    public void testJoinWithKey() throws Exception {
```
ditto

dguy reviewed (commented):
Thanks @jeyhunkarimov, i've left a few more comments

jeyhunkarimov commented:
@dguy thanks for the review. 
I tried to apply tests (with and without key) through a common method (like `testJoin(boolean withKey)`), however, I got Cyclomatic complexity error while testing. As a result, I decided to pass some complexity to another function `KStreamKstreamJoinTest.checkResult()`.
